### PR TITLE
Adopt to new luma state managment api

### PR DESCRIPTION
**Pull request by 1chandu**
- Changing gl.*** calls into setParameters
- Use withParameters for temporarily setting state for specific draws (draw-and-pick.js)
- Tested with:  'run test, run test-browser (3 errors, not new), examples/layer-browser and all website examples'.

### CODE REVIEW

File: src/lib/draw-and-pick.js [hunk 1]
```diff
@@ -354,47 +354,47 @@ function getPickedColors(gl, {
   // Make sure we clear scissor test and fbo bindings in case of exceptions
   // We are only interested in one pixel, no need to render anything else
   // Note that the callback here is called synchronously.
-  return glContextWithState(gl, {
+  return withParameters(gl, {
     frameBuffer: pickingFBO,
     framebuffer: pickingFBO,
     scissorTest: {x, y, w: width, h: height}
```
Comment (ibgreen) [inline src/lib/draw-and-pick.js]: Does this work? Didn't you decide to go with two values?

File: src/lib/draw-and-pick.js [hunk 2]
```diff
@@ -354,47 +354,47 @@ function getPickedColors(gl, {
...
-        layer.drawLayer({
-          uniforms: Object.assign(
-            {renderPickingBuffer: 1, pickingEnabled: 1},
-            layer.context.uniforms,
-            getUniformsFromViewport(layer.context.viewport, layer.props),
-            {layerIndex}
-          )
-        });
-      }
+    const settings = {
```
Comment (ibgreen) [inline src/lib/draw-and-pick.js]: Move these into the withParameters call, they should be set for all the draw calls...

File: src/lib/draw-and-pick.js [hunk 3]
```diff
@@ -354,47 +354,47 @@ function getPickedColors(gl, {
...
+      // Render all pickable layers in picking colors
+      layers.forEach((layer, layerIndex) => {
+        if (!layer.isComposite && layer.props.visible && layer.props.pickable) {
+
+          // Encode layerIndex with alpha
+          const blendColorSettings = {blendColor: [0, 0, 0, (layerIndex + 1) / 255]};
+          withParameters(gl, blendColorSettings, () => {
+            layer.drawLayer({
+              uniforms: Object.assign(
+                {renderPickingBuffer: 1, pickingEnabled: 1},
```
Comment (ibgreen) [inline src/lib/draw-and-pick.js]: For these temporary uniforms, it would be best to unset them with a setUniforms call. uniforms don't autoreset like settings. Maybe make a note about this for now.

File: src/lib/draw-and-pick.js [hunk 4]
```diff
@@ -354,47 +354,47 @@ function getPickedColors(gl, {
...
+      blendEquation: gl.FUNC_ADD
+    };
+
+    withParameters(gl, settings, () => {
+      // Render all pickable layers in picking colors
+      layers.forEach((layer, layerIndex) => {
+        if (!layer.isComposite && layer.props.visible && layer.props.pickable) {
+
+          // Encode layerIndex with alpha
+          const blendColorSettings = {blendColor: [0, 0, 0, (layerIndex + 1) / 255]};
```
Comment (ibgreen) [inline src/lib/draw-and-pick.js]: Add a `settings` parameter to layer.drawLayer and pass this in there, instead of using another nested `withParameters`.

Even just a `setParameters` call would work since we are already inside a withParameters, value will be restored on exit.

File: src/lib/layer.js [hunk 5]
```diff
@@ -419,11 +419,15 @@ export default class Layer {
...
-    gl.polygonOffset(offset[0], offset[1]);
+    const offsets = getPolygonOffset && getPolygonOffset(uniforms) || [0, 0];
+    const polygonOffsetSettings = {
+      polygonOffset: offsets
+    };
 
     // Call subclass lifecycle method
-    this.draw({uniforms});
+    withParameters(gl, polygonOffsetSettings, () => {
+      this.draw({uniforms});
```
Comment (ibgreen) [inline src/lib/layer.js]: `this.draw({uniforms, settings: polygonOffsetSettings})`

File: src/react/deckgl.js [hunk 6]
```diff
@@ -109,11 +109,11 @@ export default class DeckGL extends React.Component {
...
   _onRendererInitialized({gl, canvas}) {
-    gl.enable(GL.BLEND);
-    gl.blendFunc(GL.SRC_ALPHA, GL.ONE_MINUS_SRC_ALPHA);
-
-    // Enable polygon offset
-    gl.enable(GL.POLYGON_OFFSET_FILL);
+    setParameters(gl, {
+      blend: true,
+      blendFunc: [GL.SRC_ALPHA, GL.ONE_MINUS_SRC_ALPHA],
+      polygonOffsetFill: true
```
Comment (ibgreen) [inline src/react/deckgl.js]: Should we just move this into the draw setting you modified above, in layer.js, rather than flip it on globally?

File: src/react/webgl-renderer.js [hunk 7]
```diff
@@ -145,13 +145,7 @@ export default class WebGLRenderer extends React.Component {
...
     h = h * dpr;
 
-    if (x !== this.x || y !== this.y || w !== this.w || h !== this.h) {
-      gl.viewport(x, y, w, h);
-      this.x = x;
-      this.y = y;
-      this.w = w;
-      this.h = h;
-    }
+    setParameters(gl, {viewport: [x, y, w, h]});
```
Comment (ibgreen) [inline src/react/webgl-renderer.js]: Oh, you are relying on the internal caching, nice!

File: src/layers/core/screen-grid-layer/screen-grid-layer.js [hunk 8]
```diff
@@ -79,10 +79,13 @@ export default class ScreenGridLayer extends Layer {
   draw({uniforms}) {
     const {minColor, maxColor} = this.props;
     const {model, cellScale, maxCount} = this.state;
-    const {gl} = this.context;
-    gl.depthMask(true);
     uniforms = Object.assign({}, uniforms, {minColor, maxColor, cellScale, maxCount});
-    model.render(uniforms);
+    model.draw({
```
Comment (shaojingli) [inline src/layers/core/screen-grid-layer/screen-grid-layer.js]: @1chandu Is there a reason we start using model.draw() here? We always use model.render() in deck.gl before.

Comment (1chandu) [inline src/layers/core/screen-grid-layer/screen-grid-layer.js]: model.draw() also takes, settings and wraps render within `withParameters` call.

File: src/lib/draw-and-pick.js (hunk 2 above)
Comment (1chandu) [inline src/lib/draw-and-pick.js]: yes, I am passing settings to withParameters() below.

File: src/lib/draw-and-pick.js (hunk 4 above)
Comment (1chandu) [inline src/lib/draw-and-pick.js]: Yeah agree, given every layer only updates blendColor, I can just use setParameters, `withParameters` may be too heavy and un-necessary here.

File: src/react/deckgl.js (hunk 6 above)
Comment (1chandu) [inline src/react/deckgl.js]: I am matching the old behavior, you suggested change can go into different PR?

File: src/lib/draw-and-pick.js (hunk 1 above)
Comment (1chandu) [inline src/lib/draw-and-pick.js]: @ib didn't get your question, what are `two values` ?

File: src/lib/layer.js (hunk 5 above)
Comment (1chandu) [inline src/lib/layer.js]: Agree, will make this change.

File: src/lib/draw-and-pick.js (hunk 1 above)
Comment (ibgreen) [inline src/lib/draw-and-pick.js]: > didn't get your question, what are two values ?

`scissorTest` and `scissorBox` or something like that

Comment (1chandu) [inline src/lib/draw-and-pick.js]: Oh yeah scissorTest and scissor, this is the old code, you already fixed it :), will reflect when I update the PR.
